### ISSUE TITLE
Stratification not incorporated into cross-validation, only for train-test splitting

### ISSUE THREAD

**Issue by katy-scott** [pmcdi/jarvais#134]
# Problem
Currently, the `stratify_on` argument in the Trainer is only used for the `train_test_split` function. But it's my understanding that you would want to incorporate it during cross-validation as well for the same reasoning. For example, if a patient has two samples in the training set, you would not want them to be separated into different parts for the cross-validation. 

# Proposed Solution
To solve this, Trainer can be updated from using KFold to GroupFold.

In `autogluon_trainer.py`, the functions `fit` and `_train_autogluon_with_cv` would be updated. Both would gain an input argument called `groups` that would be used for the `GroupKFold` iterator in the cv function. 
```python
def fit(
        self,
        X_train: pd.DataFrame,
        y_train: pd.Series,
        X_test: pd.DataFrame,
        y_test: pd.Series,
        groups: pd.Series | None = None 
    ) -> tuple[TabularPredictor, pd.DataFrame, pd.Series]:

    has_test = len(X_test) > 0

    if self.k_folds > 1:
        self._predictor, X_val, y_val = self._train_autogluon_with_cv(
            X_train, 
            y_train,
            groups
        )
    ...
```

```python
def _train_autogluon_with_cv(
            self,
            X_train: pd.DataFrame,
            y_train: pd.Series,
            groups: pd.Series | None = None
        ) -> tuple[TabularPredictor, pd.DataFrame, pd.Series]:

        kf = GroupKFold(n_splits=self.k_folds)

        val_indices = []    

        for fold, (train_index, val_index) in enumerate(kf.split(X_train, y_train, groups)):
            X_train_cv, X_val_cv = X_train.iloc[train_index], X_train.iloc[val_index]
            y_train_cv, y_val_cv = y_train.iloc[train_index], y_train.iloc[val_index]
            ...
```

The input for this is already being set up in `trainer.py` in the `run` function. You can use the same `stratify_col` being set up for the train-test splitting.
```python
if self.settings.task in {'binary', 'multiclass'}:
    stratify_col = (
        y.astype(str) + '_' + self.input_data[self.settings.stratify_on].astype(str)
        if self.settings.stratify_on is not None
        else y
    )
else:
    stratify_col = None
...
# Train
self.predictor, self.X_val, self.y_val = self.trainer_module.fit(
    X_train=self.X_train, 
    y_train=self.y_train, 
    X_test=self.X_test, 
    y_test=self.y_test,
    groups=stratify_col
)
```

This makes it so that the cross-validation considers the stratification as well. 

# Limitations
* This currently works for when the test set is set to 0 and only for binary and mutliclass tasks. 
* There is no random seed argument in GroupKFold in scikit-learn 1.5.2, which autogluon depends on. scikit-learn would need to be bumped to at least 1.6.1 for this to be possible. 

# Attachments  
[trainer.py](https://github.com/user-attachments/files/30360940/trainer.py)

[autogluon_trainer.py](https://github.com/user-attachments/files/30360961/autogluon_trainer.py)
